boot/cmdline.go

```diff
@@ -134,50 +134,64 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
 	return mbl, nil
 }
 
-// ComposeRecoveryCommandLine composes the kernel command line used when booting
-// a given recovery mode system.
-func ComposeRecoveryCommandLine(model *asserts.Model, system string) (string, error) {
+const (
+	currentEdition = iota
+	candidateEdition
+)
+
+func composeCommandLine(model *asserts.Model, currentOrCandidate int, mode, system string) (string, error) {
 	if model.Grade() == asserts.ModelGradeUnset {
 		return "", nil
 	}
-
+	if mode != ModeRun && mode != ModeRecover {
+		return "", fmt.Errorf("internal error: unsupported command line mode %q", mode)
+	}
 	// get the ubuntu-seed bootloader
 	opts := &bootloader.Options{
 		NoSlashBoot: true,
-		// Looking for recovery system bootloader
-		Recovery: true,
 	}
-	mbl, err := getBootloaderManagingItsAssets(InitramfsUbuntuSeedDir, opts)
+	bootloaderRootDir := InitramfsUbuntuBootDir
+	modeArg := "snapd_recovery_mode=run"
+	systemArg := ""
+	if mode == ModeRecover {
+		// dealing with recovery system bootloader
+		opts.Recovery = true
+		bootloaderRootDir = InitramfsUbuntuSeedDir
+		// recovery mode & system command line arguments
+		modeArg = "snapd_recovery_mode=recover"
+		systemArg = fmt.Sprintf("snapd_recovery_system=%v", system)
+	}
+	mbl, err := getBootloaderManagingItsAssets(bootloaderRootDir, opts)
 	if err != nil {
 		if err == errBootConfigNotManaged {
 			return "", nil
 		}
 		return "", err
 	}
-	systemArg := fmt.Sprintf("snapd_recovery_system=%v", system)
 	// TODO:UC20: fetch extra args from gadget
 	extraArgs := ""
-	return mbl.CommandLine("snapd_recovery_mode=recover", systemArg, extraArgs)
+	if currentOrCandidate == currentEdition {
```

Comment:
When we test whether to retrieve the current or the candidate command line we may have already failed when we get the bootloader (in line 164 above). This would be the same situation that caused the previous error during install.

Reply:
Per discussion with @pedronis we will not check whether the on disk asset is managed. The *CommandLine() call is part of the interface of a managed-assets bootloader, and the boot does the call only after checking then model, which when right implies we always have a managed asset installed (otherwise we cannot fulfill what model promises). I've added a fallback for ComandLine() which assumes a well known edition (1) when the managed asset is not installed, since it matches the base edition of grub.cfg we imported into the tree.